x/ccv/provider/types/msg.go

```diff
@@ -83,167 +71,215 @@ func NewMsgAssignConsumerKey(consumerId string, providerValidatorAddress sdk.Val
-		// same behavior as in cosmos-sdk
-		panic(err)
+// ValidateBasic implements the sdk.HasValidateBasic interface.
+func (msg MsgAssignConsumerKey) ValidateBasic() error {
+	if err := validateDeprecatedChainId(msg.ChainId); err != nil {
+		return errorsmod.Wrapf(ErrInvalidMsgAssignConsumerKey, "ChainId: %s", err.Error())
 	}
-	return []sdk.AccAddress{valAddr.Bytes()}
-}
-
-// GetSignBytes returns the message bytes to sign over.
-func (msg MsgAssignConsumerKey) GetSignBytes() []byte {
-	bz := ccvtypes.ModuleCdc.MustMarshalJSON(&msg)
-	return sdk.MustSortJSON(bz)
-}
 
-// ValidateBasic implements the sdk.Msg interface.
-func (msg MsgAssignConsumerKey) ValidateBasic() error {
 	if err := ValidateConsumerId(msg.ConsumerId); err != nil {
-		return err
+		return errorsmod.Wrapf(ErrInvalidMsgAssignConsumerKey, "ConsumerId: %s", err.Error())
 	}
+
 	_, err := sdk.ValAddressFromBech32(msg.ProviderAddr)
 	if err != nil {
-		return ErrInvalidProviderAddress
+		return errorsmod.Wrapf(ErrInvalidMsgAssignConsumerKey, "ProviderAddr: %s", err.Error())
 	}
+
 	if msg.ConsumerKey == "" {
-		return ErrInvalidConsumerConsensusPubKey
+		return errorsmod.Wrapf(ErrInvalidMsgAssignConsumerKey, "ConsumerKey cannot be empty")
 	}
 	if _, _, err := ParseConsumerKeyFromJson(msg.ConsumerKey); err != nil {
-		return ErrInvalidConsumerConsensusPubKey
+		return errorsmod.Wrapf(ErrInvalidMsgAssignConsumerKey, "ConsumerKey: %s", err.Error())
 	}
+
 	return nil
 }
 
-// ParseConsumerKeyFromJson parses the consumer key from a JSON string,
-// this replaces deserializing a protobuf any.
-func ParseConsumerKeyFromJson(jsonStr string) (pkType, key string, err error) {
-	type PubKey struct {
-		Type string `json:"@type"`
-		Key  string `json:"key"`
+// ValidateBasic implements the sdk.HasValidateBasic interface.
+func (msg *MsgChangeRewardDenoms) ValidateBasic() error {
+	emptyDenomsToAdd := len(msg.DenomsToAdd) == 0
+	emptyDenomsToRemove := len(msg.DenomsToRemove) == 0
+	// Return error if both sets are empty or nil
+	if emptyDenomsToAdd && emptyDenomsToRemove {
+		return fmt.Errorf(
+			"invalid change reward denoms proposal: both denoms to add and denoms to remove are empty")
 	}
-	var pubKey PubKey
-	err = json.Unmarshal([]byte(jsonStr), &pubKey)
-	if err != nil {
-		return "", "", err
+
+	// Return error if a denom is in both sets
+	for _, denomToAdd := range msg.DenomsToAdd {
+		for _, denomToRemove := range msg.DenomsToRemove {
+			if denomToAdd == denomToRemove {
+				return fmt.Errorf(
+					"invalid change reward denoms proposal: %s cannot be both added and removed", denomToAdd)
+			}
+		}
 	}
-	return pubKey.Type, pubKey.Key, nil
-}
 
-func NewMsgSubmitConsumerMisbehaviour(submitter sdk.AccAddress, misbehaviour *ibctmtypes.Misbehaviour) (*MsgSubmitConsumerMisbehaviour, error) {
-	return &MsgSubmitConsumerMisbehaviour{Submitter: submitter.String(), Misbehaviour: misbehaviour}, nil
-}
+	// Return error if any denom is "invalid"
+	for _, denom := range msg.DenomsToAdd {
```

Comment:
nit: these validity checks can be merged in the for loops above.

Reply:
see https://github.com/cosmos/interchain-security/pull/2197/commits/e7476168172f6e36474ca9b6b59a3481a6a4468c; I also removed the nested iteration.